src/Swift.Runtime/src/Swift/Runtime/SwiftHandle.cs

```diff
@@ -90,58 +41,60 @@ public static implicit operator IntPtr(SwiftHandle value)
     /// <summary>
     /// Explicit conversion from IntPtr to SwiftHandle
     /// </summary>
-    public static explicit operator SwiftHandle(IntPtr value)
+    public static implicit operator SwiftHandle(IntPtr value)
     {
         return new SwiftHandle(value);
     }
+}
 
+/// <summary>
+/// Represents an opaque handle to a Swift object of type T.
+/// Used to manage native memory associated with a Swift object of type T.
+/// </summary>
+public sealed class SwiftSafeHandle<T> : SafeHandleZeroOrMinusOneIsInvalid where T : ISwiftObject
+{
     /// <summary>
-    /// Explicit conversion from SwiftHandle to void*
-    /// </summary>
-    public unsafe static explicit operator void*(SwiftHandle value)
-    {
-        return (void*)(IntPtr)value;
-    }
-
-    /// <summary>
-    /// Explicit conversion from void* to SwiftHandle
-    /// </summary>
-    public unsafe static explicit operator SwiftHandle(void* value)
-    {
-        return new SwiftHandle((IntPtr)value);
-    }
-
-    /// <summary>
-    /// Compare this SwiftHandle to the given object, returns true if they're the same.
-    /// </summary>
-    public override bool Equals(object? o)
-    {
-        if (o is SwiftHandle nh)
-            return nh.handle == this.handle;
-        return false;
-    }
-
-    /// <summary>
-    /// Generate a hashcode for the SwiftHandle
+    /// Returns a SwiftSafeHandle with a zero value
     /// </summary>
-    public override int GetHashCode()
-    {
-        return handle.GetHashCode();
-    }
+    public readonly static SwiftSafeHandle<T> Zero = new SwiftSafeHandle<T>(IntPtr.Zero);
 
     /// <summary>
-    /// Returns true if this matches the give SwiftHandle
+    /// Constructs a SwiftSafeHandle from the given IntPtr
     /// </summary>
-    public bool Equals(SwiftHandle other)
+    public SwiftSafeHandle(IntPtr handle)
+        : base(ownsHandle: true)
     {
-        return other.handle == handle;
+        SetHandle(handle);
     }
 
     /// <summary>
-    /// Generates a string representation of this SwiftHandle
+    /// Releases the handle to the Swift object
     /// </summary>
-    public override string ToString()
+    protected override unsafe bool ReleaseHandle()
     {
-        return "0x" + handle.ToString("x");
+        bool success = false;
+        try
+        {
+            TypeMetadata metadata = SwiftObjectHelper<T>.GetTypeMetadata();
+            metadata.ValueWitnessTable->Destroy((void*)handle, metadata);
+            success = true;
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine($"Error releasing handle: {ex}");
+        }
+
+        try
+        {
+            NativeMemory.Free((void*)handle);
```

Comment:
`NativeMemory.Free` won't ever throw an exception. And if does throw an exception due to things being horribly wrong, you do not want to catch it.

Reply:
Thanks, updated.